contracts/competition/Competition.sol

```diff
@@ -0,0 +1,138 @@
+
+    modifier isNotMember()
+    {
+        require(membersToTeam[msg.sender] == address(0), "Team members are not allowed.");
+        _;
+    }
+
+    constructor (
+        uint _start,
+        uint _end,
+        uint _registrationStart,
+        uint _registrationEnd,
+        IReferralStorage _referralStorage
+    ) public
+    {
+        start = _start;
+        end = _end;
+        registrationStart = _registrationStart;
+        registrationEnd = _registrationEnd;
+        referralStorage = _referralStorage;
+    }
+
+    function registerTeam (string calldata name, bytes32 referral) external registrationIsOpen isNotLeader
+    {
+        require(referralStorage.codeOwners(referral) != address(0), "Referral code does not exist.");
+        require(teamNames[name] == false, "Team name already registered.");
+
+        Team storage team;
+
+        team.leader = msg.sender;
+        team.name = name;
+        team.referral = referral;
+        team.members.push(msg.sender);
+
+        teams[msg.sender] = team;
+        leaders.push(msg.sender);
+        teamNames[name] = true;
+    }
+
+    function createJoinRequest (address leaderAddress) external registrationIsOpen isNotLeader isNotMember
+    {
+        require(membersToTeam[msg.sender] == address(0), "You can't join multiple teams.");
+        require(teams[msg.sender].leader != address(0), "The team does not exist.");
+        require(requests[leaderAddress][msg.sender] == false, "You already applied for this team.");
+
+        teams[leaderAddress].joinRequests.push(msg.sender);
+        requests[leaderAddress][msg.sender] = true;
+    }
+
+    function approveJoinRequest (address memberAddress) external registrationIsOpen isNotMember
+    {
+        require(requests[msg.sender][memberAddress] == false, "This member did not apply.");
+        require(membersToTeam[memberAddress] == address(0), "This member already joined a team.");
+
+        referralStorage.setTraderReferralCode(memberAddress, teams[msg.sender].referral);
+        teams[msg.sender].members.push(msg.sender);
+        membersToTeam[memberAddress] = msg.sender;
+    }
+
+    function getLeaders() external view returns (address[] memory)
+    {
+        address[] memory res;
+
+        for (uint i = 0; i < leaders.length; i++) {
+            res[i] = leaders[i];
+        }
+
+        return res;
+    }
+
+    function getTeam(address leaderAddr) external view returns (address leader, string memory name, bytes32 referral)
+    {
+        Team memory team = teams[leaderAddr];
+        return (team.leader, team.name, team.referral);
+    }
+
+    function getMemberTeam(address memberAddr) external view returns (address)
+    {
+        return membersToTeam[memberAddr];
+    }
+
+    function getTeamMembers(address leaderAddr) external view returns (address[] memory)
+    {
+        return teams[leaderAddr].members;
+    }
+
+    function getTeamJoinRequests(address leaderAddr) external view returns (address[] memory)
```

Comment:
this is to allow leader to view all pending requests, right?

Reply:
Yes, I wanted to display all pending join requests on the team page (only if the connected account is the leader account ofc)

Reply:
i'm thinking if somebody will "spam" with lots of join requests and leader will see all those spammy requests. would be kind of a weird "attack" though :)

Reply:
Good take, wouldn't be surprised if this happens tbh

Reply:
at first glance it sounds okay to me 🤔
- traders sends request to join
- and sends address to leader
- leader enters that address and submits it

afaik there won't be too many traders? and team registration is a one-time action. does it sound reasonable?